File: 0648-replace-words/0648-replace-words.py
class TrieNode:
    children = []
    isEnd = False
    word = ""

    def __init__(self):
        self.children = [None] * 26
        self.isEnd = False
        self.word = ""

class Solution:
    def __init__(self):
        self.root = TrieNode()

    def replaceWords(self, dictionary: List[str], sentence: str) -> str:
        # if len(sentence) == 0 or sentence is None or len(dictionary) == 0 or dictionary is None:
        #     return sentence
        
        sentence = sentence.split(" ")

        for word in dictionary:
            self.insert(word, self.root)
        
        result = []

        for j in range(len(sentence)):
            if j != 0:
                result.append(" ")

            word = sentence[j]
            replacement = []
            curr = self.root

            for i in range(len(word)):
                ch = word[i]
                
                if not curr.children[ord(ch) - ord("a")] or curr.isEnd:
                    break

                replacement.append(ch)
                curr = curr.children[ord(ch) - ord("a")]
                
            if curr.isEnd:
                result.append("".join(replacement))
            else:
                result.append(word)

        
        return "".join(result)                

    
    def insert(self, word: str, curr: TrieNode) -> None:
        print(word)
        for ch in word:
            if not curr.children[ord(ch) - ord("a")]:
                curr.children[ord(ch) - ord("a")] = TrieNode()
            
            curr = curr.children[ord(ch) - ord("a")]
        
        curr.isEnd = True
        curr.word = word

File: 0648-replace-words/0648-replace-words.java
class TrieNode{
    TrieNode[] children;
    boolean isEnd;
    String currWord;
    public TrieNode(){
        this.children = new TrieNode[26];
        this.isEnd = false;
        currWord = "";
    }
}
class Solution {
    TrieNode root;

    public Solution(){
        this.root = new TrieNode();
    }
    public String replaceWords(List<String> dictionary, String sentences) {

        for(String word: dictionary){
            insert(word, root);
        }
        StringBuilder result = new StringBuilder();

        String[] sentence = sentences.split(" ");

        for(int j = 0; j < sentence.length; j++){
            if(j != 0){
                result.append(" ");
            }

            String word = sentence[j];
            StringBuilder replacement = new StringBuilder();
            TrieNode curr = root;

            for(int i = 0; i < word.length(); i++){
                char ch = word.charAt(i);

                if(curr.children[ch - 'a'] == null || curr.isEnd == true){
                    break;
                }

                replacement.append(ch);
                curr = curr.children[ch - 'a'];
                
            }
            if(curr.isEnd == true){
                result.append(replacement.toString());
            }else{
                result.append(word);
            }
        }

        return result.toString();
    }


    private void insert(String word, TrieNode curr){
        
        for(int i = 0; i < word.length(); i++){
            char ch = word.charAt(i);

            if(curr.children[ch - 'a'] == null){
                curr.children[ch - 'a'] = new TrieNode();
            }

            curr = curr.children[ch - 'a'];
        }

        curr.isEnd = true;
        curr.currWord = word;
    }
}


// TrieNode
//